#798 - PackageReivsion deletionProposed sometimes deletes the packageRevision

kushnaidu opened the pull request:
resolves https://github.com/nephio-project/nephio/issues/798 
This change sends a notification of packageRevision change before creating packageRev to avoid the race condition of an invalid ownerRef uid for packageRev, which causes k8s to add a deletionTimestamp to the packageRevs.

kushnaidu commented:
/retest

kispaljr commented:
I may be wrong, but it seems to me that by deleting those lines from code you removed two functionality from porch:
- OwnerReferences set to a PackageRevision object (e.g. by a PackageVariant) are no longer stored in the metadata store. That means if e.g. the porch server is restarted, those OwnerReferences are lost forever (preventing this is the reason why we have a metadata store in the first place)
- The internal metadata store CR (kind: PackageRev) doesn't have an ownerReference to its parent PackageRevsision object any more. That ownerReference also existed for a reason, among others to let the garbage collector know that if the PackageRevision is deleted it can also delete the PackageRev.

Could you elaborate  on the reasons and implications of removing these functionalities?

kushnaidu commented:
> I may be wrong, but it seems to me that by deleting those lines from the code you removed the following functionality from porch: The internal metadata store CR (kind: PackageRev) doesn't have an ownerReference to its parent PackageRevision object any more. That ownerReference exists mainly to let the garbage collector know that if the PackageRevision is deleted it can also delete the PackageRev.
> 
> Could you elaborate on the reasons and implications of removing this functionality?

Hey @kispaljr, thanks for your feedback. 
The issue is that after an approve, when the 'main' twin gets created, intermittently, k8s adds the deletionTimestamp to it. 
The suspect seemed to be the ownerRefs. Sometimes k8s sees the ownerRefs before the pkgrev is created and it looks at the pkgrev as an orphan. So upon a propose-delete, porch checks if the pkgRev has a deletionTimestamp and if it does, it removes the finalizer and calls for a delete of the pkgrev.

kispaljr commented:
Thank you for the clarification. I think I understand it much better now.
Can we find a solution that keeps the ownerReference in the PackageRev pointing to the PackageRevision, but avoid the race condition at the same time?

kushnaidu commented:
In the latest commit, I've reverted the removal of the ownerRefs and changed the flow in repo.go to send a notification of the packagerevision change first before creation of the packageRev.

liamfallon reviewed (commented):
Looks good

/approve

efiacor commented on pkg/cache/repository.go:
```diff
@@ -472,25 +472,8 @@ func (r *cachedRepository) refreshAllCachedPackages(ctx context.Context) (map[re
-	}
-
 	// Send notification for packages that changed.
+	// Send notification first before creation of PkgRev to avoid race condition after an approve because of ownerReference.
```
Can we merge these 2 comments into 1?

kushnaidu replied:
sure, I can do that.

kushnaidu replied:
Just made the change.

efiacor reviewed (commented):
Minor nit pick

kushnaidu commented on pkg/cache/repository.go:
```diff
@@ -472,25 +472,8 @@ func (r *cachedRepository) refreshAllCachedPackages(ctx context.Context) (map[re
-		}
-	}
-
 	// Send notification for packages that changed.
+	// Send notification first before creation of PkgRev to avoid race condition after an approve because of ownerReference.
```
```suggestion
	// Send notification for packages that changed before creation of PkgRev to avoid race condition after an approve because of ownerReferences.
```
```suggestion
	// Send notification for packages that changed.
	// Send notification first before creation of PkgRev to avoid race condition after an approve because of ownerReference.
```
```suggestion
	// Send notification for packages that changed before the creation of PkgRev to avoid race conditions because of ownerReferences.
```

kushnaidu replied:
```suggestion
	// Send notification for packages that changed before the creation of PkgRev to avoid race conditions because of ownerReferences.
```

Catalin-Stratulat-Ericsson commented:
/approve

efiacor commented:
/approve

liamfallon commented:
@kispaljr does the later changes address your concerns?

kispaljr commented:
sorry, I haven't got the time to review the new version till now.
The current implementation is definitely an improvement, so I support merging it.

I still cannot see however, how this change guarantees that the parent PackageRevision exists in the porch server by the time the PackageRev is created in the etcd. Please note, that I am not saying that it doesn't guarantee it, I just don't see why, and I am afraid that it only decreases the probability of the race condition to happen by delaying the creation of PackageRevs with the time it takes to send out the notifications.
I guess all I am trying to say is that an explanation of how this solves the problem should be put in a comment in the code, so that developers would understand it easier.

kispaljr commented:
Since as I stated there is no doubt that this is an improvement, and the solution was tested so we are sure that in practice it solves the original issue, please ignore my theoretical thoughts above and let's merge this.

/lgtm
/approve

kushnaidu commented:
> Merge

Hey Istvan,
I tested this change with around 150-200 package revisions and couldn't reproduce the issue even once. When the cached repo is updated, it gives enough time for the pkgRev to use a valid uid I think.

nagygergo commented:
Yeah, the race condition doesn't fully disappear. If the kube-controller-manager is processing through notifications from porch much slower than processing through notifiactions from kube-apiserver's pkgrev watch, the same issue can happen. 

But this can happen between any 2 resources. I'll check whether there is a safeguard against this in the garbage collector.
